File: 15.memento/memento.py
# Memento
class TextEditorMemento:
    def __init__(self, state):
        self.state = state

    def get_state(self):
        return self.state

# Originador
class TextEditor:
    def __init__(self):
        self.content = ""

    def set_content(self, content):
        self.content = content

    def get_content(self):
        return self.content

    def save(self):
        return TextEditorMemento(self.content)

    def restore(self, memento):
        self.content = memento.get_state()

# Cuidador
class Caretaker:
    def __init__(self):
        self.mementos = []

    def add_memento(self, memento):
        self.mementos.append(memento)

    def get_memento(self, index):
        return self.mementos[index]

# Uso
editor = TextEditor()
caretaker = Caretaker()

editor.set_content("Estado 1")
caretaker.add_memento(editor.save())

editor.set_content("Estado 2")
caretaker.add_memento(editor.save())

editor.set_content("Estado 3")

print("Estado actual:", editor.get_content())
editor.restore(caretaker.get_memento(0))
print("Estado restaurado:", editor.get_content())

File: 15.memento/memento.java
// Memento
class TextEditorMemento {
    private final String state;

    public TextEditorMemento(String state) {
        this.state = state;
    }

    public String getState() {
        return state;
    }
}

// Originador
class TextEditor {
    private String content;

    public void setContent(String content) {
        this.content = content;
    }

    public String getContent() {
        return content;
    }

    public TextEditorMemento save() {
        return new TextEditorMemento(content);
    }

    public void restore(TextEditorMemento memento) {
        content = memento.getState();
    }
}

// Cuidador
class Caretaker {
    private final List<TextEditorMemento> mementos = new ArrayList<>();

    public void addMemento(TextEditorMemento memento) {
        mementos.add(memento);
    }

    public TextEditorMemento getMemento(int index) {
        return mementos.get(index);
    }
}

// Uso
public class memento {
    public static void main(String[] args) {
        TextEditor editor = new TextEditor();
        Caretaker caretaker = new Caretaker();

        editor.setContent("Estado 1");
        caretaker.addMemento(editor.save());

        editor.setContent("Estado 2");
        caretaker.addMemento(editor.save());

        editor.setContent("Estado 3");

        System.out.println("Estado actual: " + editor.getContent());
        editor.restore(caretaker.getMemento(0));
        System.out.println("Estado restaurado: " + editor.getContent());
    }
}